dot/network/discovery.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2019 ChainSafe Systems (ON) Corp.
+// This file is part of gossamer.
+//
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	badger "github.com/ipfs/go-ds-badger2"
+	libp2phost "github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	libp2pdiscovery "github.com/libp2p/go-libp2p-discovery"
+	kaddht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/libp2p/go-libp2p-kad-dht/dual"
+)
+
+// discovery handles discovery of new peers via the kademlia DHT
+type discovery struct {
+	ctx                context.Context
+	dht                *dual.DHT
+	h                  libp2phost.Host
+	bootnodes          []peer.AddrInfo
+	ds                 *badger.Datastore
+	pid                protocol.ID
+	minPeers, maxPeers int
+}
+
+func newDiscovery(ctx context.Context, h libp2phost.Host, bootnodes []peer.AddrInfo, ds *badger.Datastore, pid protocol.ID, min, max int) *discovery {
+	return &discovery{
+		ctx:       ctx,
+		h:         h,
+		bootnodes: bootnodes,
+		ds:        ds,
+		pid:       pid,
+		minPeers:  min,
+		maxPeers:  max,
+	}
+}
+
+// start creates the DHT.
+func (d *discovery) start() error {
+	if len(d.bootnodes) == 0 {
+		// get all currently connected peers and use them to bootstrap the DHT
+		peers := d.h.Network().Peers()
+
+		for {
+			if len(peers) == 0 {
+				logger.Info("no peers yet, waiting to start DHT...")
+				time.Sleep(time.Second * 10) // wait for peers to connect before starting DHT
```

Comment:
maybe this `time.Second * 10` could be a constant?

Reply:
yes I will update all these constants before un-drafting the PR